### PR TITLE
Add burn context: context-file cost attribution across CLAUDE.md + AGENTS.md

### LINKED ISSUE #10: CLAUDE.md hot-path cost attribution with section-level recommendations
## Context

Prism (jakeefr/prism) validates this as the single highest-leverage waste-finder in a Claude Code project: CLAUDE.md rides in every turn's cached context, so a bloated CLAUDE.md silently taxes every session in the repo forever. Their real-data examples (from the README):

- *\"6738% CLAUDE.md re-read cost in one session: a 237-line file being re-read on every tool call\"*
- *\"CLAUDE.md re-reads consumed 480% of total session tokens\"*

Prism's mechanism is a proxy: `reread_cost = tool_call_count × claude_md_size_tokens` (`analyzer.py:173`). It never touches real `cacheRead` values. The shock ratios are `(count × size) / session_incremental_tokens`, meant to alarm, not to be precise.

Burn can do this correctly because the ledger already carries real per-turn `cacheRead` values.

## Scope

One file (plus nested variants): the project's `CLAUDE.md` at repo root. Also detect nested `.claude/CLAUDE.md` or scoped `src/**/CLAUDE.md` files Claude Code loads hierarchically.

### Cost attribution

```
At session start, resolve the active CLAUDE.md set from cwd / project root.
Measure byte length of each file; estimate tokens (bytes/4 heuristic, or tiktoken cl100k if the dep cost is acceptable).

For each turn T in the session:
  share_T = claude_md_tokens / (claude_md_tokens + tool_defs_tokens + conversation_tokens_T)
  attributed_tokens_T = usage.cacheRead_T × share_T
  attributed_cost_T   = attributed_tokens_T × cache_read_price(model)

session_claude_md_cost = Σ attributed_cost_T
```

- `tool_defs_tokens` — sum of the registered tools' schema tokens. Approximate once per session from the Claude Code client version or derive from the session JSONL if exposed; otherwise treat as a fixed constant per Claude Code minor version and measure once.
- `conversation_tokens_T` — back-compute from `usage.input - cacheCreate - cacheRead` plus the running prefix. Rough is fine; we're computing a share, not a precise boundary.
- Accounts for cache-read pricing being ~10% of fresh-input price (prism doesn't — they multiply tokens by tool-call count as if every read was fresh).

### Section ranking

Unlike prism's zone heuristic (top 20%, mid 20-75%, bottom 25% at `advisor.py:234-235`), parse CLAUDE.md by **markdown heading structure**:

- `# Top-level` / `## Section` / `### Subsection` — each heading begins a section.
- Capture each section's line range, heading text, and token count.
- Compute per-section cost as `(section_tokens / total_claude_md_tokens) × session_claude_md_cost`.

No keyword classification (prism does `tone|style|personality` regex — too fragile). Rank purely by cost. The section *content* is what the user judges, not a classifier-assigned category.

## CLI surface

```
burn claude-md [--project <path>] [--since 7d]
# Prints:
#   CLAUDE.md at /Users/will/Projects/foo (412 lines, ~3.1k tokens)
#   Cost per session:   avg $0.18, p95 $0.34
#   Cost last 7 days:   $4.12 across 23 sessions
#
#   Sections ranked by cost:
#     lines  12- 47  ## Architecture              1,820 tok  $0.11/session
#     lines  48- 89  ## Testing conventions         890 tok  $0.05/session
#     lines  90-148  ## Tone and personality notes  410 tok  $0.02/session   <- consider trimming
#     …

burn claude-md advise [--project <path>]
# Emits a unified diff of suggested TRIM / RESTRUCTURE hunks with
# per-recommendation projected-savings-per-session and savings-over-last-N-sessions.
# Does NOT write to the file. User applies manually.

burn claude-md --json   # programmatic
```

Explicitly **no** `--apply` flag. Unlike prism's `advise --apply`, burn never mutates files.

## Acceptance

- Attribution math: on a fixture session with a known CLAUDE.md and known `cacheRead` values, `burn claude-md` reports a per-turn attributed cost within ±10% of a hand-computed baseline.
- Section parsing correctly handles: top-level text before the first heading (treated as a \"preamble\" section), nested headings (rolled up to their parent level for ranking), and files with no headings (reported as a single section with a note).
- Cross-session view: `burn claude-md --since 7d` aggregates across sessions for the same `projectKey` (depends on #4's git canonicalization).
- `advise` output is applyable by hand: line ranges match the user's current file (run against HEAD, not a stale snapshot).
- If CLAUDE.md has changed during the reporting window, attribution uses the CLAUDE.md *at the time of each session*, not the current version. (Read via git log if the file is tracked; fall back to current file with a warning otherwise.)

## Depends on

- #4 (git-canonical projectKey) for cross-session rollup.
- Nothing else strictly blocking. Can land before #3 (`burn waste`) — this is the most user-visible waste finder.

## Unblocks

The biggest concrete \"switch a choice, save spend\" answer for the meta-goal. Higher leverage than model-switching because CLAUDE.md savings compound across every future session in the repo.

## Out of scope

- Automated CLAUDE.md rewriting (`advise --apply` equivalent) — deliberate; burn recommends, doesn't mutate.
- Adherence / \"Mr. Tinkleberry\" detection — prism does this with hand-coded per-rule checkers (`analyzer.py:548+`). Narrow and doesn't scale. See #6.

### PR DESCRIPTION
**Pull request by willwashburn**
## Summary

- **`burn context`** — cost attribution for agent context files across every agent `burn` ingests. Discovers `CLAUDE.md`, `.claude/CLAUDE.md` (Claude Code), and `AGENTS.md` (Codex + OpenCode). Per-file ranked section tables plus a grand total across all files. No cross-attribution: a Codex session never pays for CLAUDE.md, a Claude Code session never pays for AGENTS.md.
- **`burn context advise`** — unified-diff TRIM hunks for the most expensive sections across every discovered file. Read-only — burn never mutates your context files.
- **`--kind claude-md | agents-md`** narrows either command to one file kind when you only care about one.

Closes #10.

### `burn context`

\`\`\`
Context files in /Users/will/Projects/foo:

CLAUDE.md (/Users/will/Projects/foo/CLAUDE.md) — 412 lines, ~3.1k tokens — applies to: claude-code
  Cost per session:   avg \$0.18, p95 \$0.34
  Cost over 7d: \$4.12 across 23 sessions
  Sections ranked by cost:
    lines   heading                    tokens  cost/session  %file
      12- 47  ## Architecture          1.8k    \$0.11         59.0%
      48- 89  ## Testing conventions   890     \$0.05         28.7%
      …

AGENTS.md (/Users/will/Projects/foo/AGENTS.md) — 95 lines, ~800 tokens — applies to: codex, opencode
  Cost per session:   avg \$0.04, p95 \$0.07
  Cost over 7d: \$0.18 across 8 sessions
  Sections ranked by cost:
    …

Grand total (all context files, 7d): \$4.30
\`\`\`

### `burn context --kind claude-md`

Same output but narrowed to CLAUDE.md files only.

## Attribution math

Every turn's cost for a context file is computed directly: \`file_tokens × cacheReadPrice / 1M\` per turn whose \`cacheRead >= file_tokens\` **and** whose \`source\` is in the file's \`appliesTo\` list. The cache-size check skips:

- The first turn of a session (the file lives in \`cacheCreate\`, not \`cacheRead\`)
- Turns where the file has been compacted out of cache

Section attribution is proportional to bytes (strictly additive), so \`Σ sectionCost ≤ totalCost\` always.

## Section parsing

- Groups by **H2** if any exist (matches the standard CLAUDE.md / AGENTS.md shape: H1 doc title + H2 sections)
- Falls back to H1 if no H2s
- Treats a heading-less file as a single \"preamble\" section
- Strict CommonMark fenced-code-block handling: only a run of the same fence char (≥ opening length) + whitespace closes
- CRLF → LF normalization and trailing-newline stripping so line numbers match what an editor shows

## Test plan

- [x] \`pnpm run build\` — clean
- [x] \`pnpm run test\` — 204/204 passing
- [x] Synthetic-attribution test (5 turns × known file size) within ±10% of hand-computed truth
- [x] Multi-harness routing test: Claude Code turn routes to CLAUDE.md, Codex/OpenCode turns route to AGENTS.md
- [x] No cross-attribution test: a Codex-only session with a CLAUDE.md present attributes \$0 to CLAUDE.md
- [x] Section parser tests: preamble, H2 grouping, H1 fallback, code-fence skip, CRLF, trailing newline, empty input
- [x] Additivity test: 20 sections, Σ share = 1.0 exactly
- [x] Advise diff test: hunk ranges match the file; POSIX-relative path when baseDir is given
- [x] \`findContextFiles\` detects all three files with correct \`appliesTo\`
- [x] CLI help lists \`context\` with subcommands and --kind flag

## Notes / follow-ups

- Per-session-time-travel (file as it existed at session time, via git log) is out of scope for MVP — always uses HEAD.
- Nested \`src/**/CLAUDE.md\` discovery not yet implemented; only root and \`.claude/CLAUDE.md\` are scanned.
- User-global context files (\`~/.claude/CLAUDE.md\`, \`~/.codex/AGENTS.md\`) not scanned — project-local only. Easy to add as a separate pass.
- Rollup across worktrees via \`projectKey\` is wired.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by barryollama — APPROVED** [review]
LGTM! This is a well-implemented feature for CLAUDE.md cost attribution.

**Highlights:**
- Clean section parsing logic with H2-first grouping, falling back to H1
- Smart handling of code fences to exclude example headings
- Conservative cost attribution (skips when cacheRead < CLAUDE.md size)
- Comprehensive test coverage (11 tests, 254 lines)
- Useful  subcommand with unified-diff TRIM hunks

The proportional attribution math and session-level aggregation look correct. Test coverage for edge cases (empty CLAUDE.md, first turn in cacheCreate, compacted cache) is solid.

One minor note for future consideration: The  estimate is reasonable for English text, but could drift for content with heavy punctuation or non-Latin scripts. Not blocking for MVP.

🦙 Approved

**Review by barryollama — APPROVED** [review]
LGTM! This is a well-implemented feature for CLAUDE.md cost attribution.

Highlights:
- Clean section parsing with H2-first grouping, H1 fallback
- Smart code fence handling to exclude example headings
- Conservative cost attribution (skips when cacheRead is below CLAUDE.md size)
- Comprehensive tests: 11 new tests, all 193 passing
- Useful advise subcommand with unified-diff TRIM hunks

The proportional attribution math looks correct. Edge case coverage (empty file, cacheCreate turns, compacted cache) is solid.

One note: The chars-per-token estimate works well for English but could drift with heavy punctuation or non-Latin scripts. Acceptable for MVP.

Approved 🦙

File: packages/analyze/src/claude-md.ts [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+export function parseClaudeMd(filePath: string, text: string): ParsedClaudeMd {
+  const lines = text.split('\n');
+  const totalLines = lines.length;
+  const bytes = Buffer.byteLength(text, 'utf8');
+  const tokens = estimateTokens(text);
+
+  const headings = findHeadings(lines);
+  // Choose the grouping level:
+  //   - If any H2 exists: group at H2 (per the issue's example output).
+  //   - Else if any H1 exists: group at H1.
```
Comment (barryollama) [inline packages/analyze/src/claude-md.ts:121]: **Nit:** The fence matching logic here uses `startsWith(marker.charAt(0).repeat(marker.length))` which checks that the closing fence contains at least the same number of fence characters as the opening.

While this works for standard cases, consider that a fence opened with 6 backticks and a line like 'backtick x 5 + text' would still match. CommonMark specifies closing fences should have at least as many backticks as the opening and should be followed only by whitespace.

Not a blocker since this is primarily for CLAUDE.md which is controlled content, but worth considering stricter validation.

**Review by barryollama — COMMENTED** [review]
LGTM with one minor suggestion about fence parsing edge case handling.

**Review by barryollama — APPROVED** [review]

File: packages/analyze/src/claude-md.ts [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    const trimmed = line.trim();
+    // Skip headings inside fenced code blocks. Match opening/closing fences
+    // (``` or ~~~, possibly followed by a language tag).
+    const fenceMatch = trimmed.match(/^(```+|~~~+)/);
+    if (fenceMatch) {
+      const marker = fenceMatch[1]!;
+      if (!inFence) {
+        inFence = true;
+        fenceMarker = marker;
+      } else if (trimmed.startsWith(fenceMarker.charAt(0).repeat(fenceMarker.length))) {
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/analyze/src/claude-md.ts]: Fixed in 97c96db. The fence-close regex now requires a run of only the opening fence char (with length ≥ opening run) followed by optional whitespace — so a line like `\`\`\`\`python` inside a 3-backtick block no longer closes the fence. Added a test that exercises the exact reproduction from the review.

File: packages/analyze/src/claude-md.ts (hunk 1 above)
Comment (willwashburn) [reply packages/analyze/src/claude-md.ts:121]: Good catch — fixed in 97c96db. Switched to a strict CommonMark close check (same fence char, ≥ opening run length, only whitespace after). Added a regression test.

File: packages/analyze/src/claude-md.ts [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+      const dominantModel = pickDominantModel(modelCounts);
+      sessionCosts.push({
+        sessionId,
+        cost,
+        ridingTurns,
+        totalTurns: turns.length,
+        model: dominantModel,
+      });
+      totalCost += cost;
+    }
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/analyze/src/claude-md.ts]: Fixed in 97c96db. Sessions with zero attributed cost are now recorded too (with cost=0, ridingTurns=0), so sessionCount/perSessionAvg/perSessionP95 reflect the whole query window. Added a test that covers a mixed (cached + uncached) two-session case.

File: packages/cli/src/commands/claude-md.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+}
+
+function formatTokens(tokens: number): string {
+  if (tokens >= 1000) return `${(tokens / 1000).toFixed(1)}k`;
+  return String(tokens);
+}
+
+function resolveProjectPath(args: ParsedArgs): string {
+  const flag = args.flags['project'];
+  if (typeof flag === 'string' && flag.length > 0) return path.resolve(flag);
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/cli/src/commands/claude-md.ts]: Fixed in 97c96db. The filesystem path is now only used for CLAUDE.md file discovery; the ledger query uses the git-canonical projectKey when available and falls back to the path when there is no git remote. No more unconditional `path.resolve` on a value that could be a projectKey.

File: packages/cli/src/commands/claude-md.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    return null;
+  }
+
+  const q: Query = { project: projectPath };
+  if (typeof args.flags['since'] === 'string') q.since = parseSinceArg(args.flags['since']);
+
+  await ingestAll();
+  const pricing = await loadPricing();
+  const turns = await queryAll(q);
+  const attribution = attributeClaudeMd({ files, turns, pricing });
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/cli/src/commands/claude-md.ts]: Fixed in 97c96db — I now call `resolveProject(projectPath)` from `@relayburn/reader` and pass its `projectKey` as `q.project` when resolvable. Since `Query.project` matches on either `turn.project` or `turn.projectKey`, this rolls up across worktrees of the same repo.

File: packages/analyze/src/claude-md.ts [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+export function attributeClaudeMd(
+  input: AttributeClaudeMdInput,
+): ClaudeMdAttributionResult {
+  const totalTokens = input.files.reduce((sum, f) => sum + f.tokens, 0);
+  if (totalTokens === 0) {
+    return {
+      totalTokens: 0,
+      totalCost: 0,
+      sessionCosts: [],
+      sectionCosts: [],
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/analyze/src/claude-md.ts:233]: Fixed in 97c96db. Section bytes are now computed additively (Σ section.bytes == totalBytes exactly), via a per-line weight table that includes each lines trailing \n. Shares are computed from byte shares rather than rounded-up token counts, so Σ sectionCost ≤ totalCost always. Added a regression test with 20 sections asserting Σ share == 1.0 exactly.

File: packages/analyze/src/claude-md.ts [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+  const sessionCostValues = sessionCosts.map((s) => s.cost).sort((a, b) => a - b);
+  const perSessionAvg = sessionCostValues.length === 0
+    ? 0
+    : sessionCostValues.reduce((a, b) => a + b, 0) / sessionCostValues.length;
+  const perSessionP95 = percentile(sessionCostValues, 0.95);
+
+  const sectionCosts: SectionCost[] = [];
+  for (const f of input.files) {
+    for (const section of f.sections) {
+      const tokenShare = section.tokens / totalTokens;
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/analyze/src/claude-md.ts]: Addressed in 97c96db together with the other additivity comment — tokenShare is now computed from section.bytes / totalBytes, which is strictly additive.

File: packages/analyze/src/claude-md.ts [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+  return found;
+}
+
+export async function loadClaudeMdFile(filePath: string): Promise<ParsedClaudeMd> {
+  const text = await readFile(filePath, 'utf8');
+  return parseClaudeMd(filePath, text);
+}
+
+export function parseClaudeMd(filePath: string, text: string): ParsedClaudeMd {
+  const lines = text.split('\n');
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/analyze/src/claude-md.ts]: Fixed in 97c96db. parseClaudeMd now normalizes CRLF → LF, drops a single trailing newline so totalLines matches what a user sees in an editor, and handles empty input (returns 0 lines). Added tests for trailing-newline and CRLF cases.

File: packages/analyze/src/claude-md.ts [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+  filePath: string,
+  fileText: string,
+  rec: AdviseRecommendation,
+): string {
+  const lines = fileText.split('\n');
+  const start = rec.section.startLine;
+  const end = rec.section.endLine;
+  const removed = lines.slice(start - 1, end);
+  // Use the path inside the project for relative display.
+  const display = filePath;
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/analyze/src/claude-md.ts]: Fixed in 97c96db. renderUnifiedDiffForRecommendation now takes an optional baseDir and emits a POSIX-normalized relative path (no leading slash) in the diff header. The CLI passes projectPath as baseDir. Added a test asserting `--- a/CLAUDE.md` rather than `--- a//abs/path/CLAUDE.md`.

File: packages/analyze/src/claude-md.test.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  });
+
+  it('finds root CLAUDE.md and .claude/CLAUDE.md', async () => {
+    await writeFile(path.join(tmp, 'CLAUDE.md'), '# Root');
+    await mkdir(path.join(tmp, '.claude'), { recursive: true });
+    await writeFile(path.join(tmp, '.claude', 'CLAUDE.md'), '# Nested');
+    const files = await findClaudeMdFiles(tmp);
+    assert.equal(files.length, 2);
+    assert.ok(files.some((f) => f.endsWith('/CLAUDE.md') && !f.includes('.claude')));
+    assert.ok(files.some((f) => f.endsWith('.claude/CLAUDE.md')));
```
[the remark this answers is not in the document]
Comment (willwashburn) [reply packages/analyze/src/claude-md.test.ts]: Fixed in 97c96db. The test now uses path.basename/path.dirname instead of endsWith with hard-coded POSIX separators, so it works on Windows too.
